ddtrace/internal/import_hooks.py

```diff
@@ -0,0 +1,360 @@
+
+    return exec_and_call_hooks(module_name, wrapped, args, kwargs)
+
+
+def wrapped_find_and_load_unlocked(wrapped, instance, args, kwargs):
+    """
+    Wrapper for `importlib._bootstrap._find_and_load_unlocked` so we can trigger
+      hooks on module loading
+
+    NOTE: This code does not get called for module reloading
+    """
+    module_name = None
+    try:
+        module_name = args[0]
+    except Exception:
+        log.debug('Failed to determine module name when importing module: %r', args, exc_info=True)
+        return wrapped(*args, **kwargs)
+
+    return exec_and_call_hooks(module_name, wrapped, args, kwargs)
+
+
+def wrapped_import(wrapped, instance, args, kwargs):
+    """
+    Wrapper for `__import__` so we can trigger hooks on module loading
+    """
+    module_name = None
+    try:
+        module_name = args[0]
+    except Exception:
+        log.debug('Failed to determine module name when importing module: %r', args, exc_info=True)
+        return wrapped(*args, **kwargs)
+
+    # Do not call the hooks every time `import <module>` is called,
+    #   only on the first time it is loaded
+    if module_name and module_name not in sys.modules:
+        return exec_and_call_hooks(module_name, wrapped, args, kwargs)
+
+    return wrapped(*args, **kwargs)
+
+
+# Keep track of whether we have patched or not
+_patched = False
+
+
+def _patch():
+    # Only patch once
+    global _patched
+    if _patched:
+        return
+
+    # 3.x
+    if PY3:
+        # 3.4: https://github.com/python/cpython/blob/3.4/Lib/importlib/_bootstrap.py#L2207-L2231
+        # 3.5: https://github.com/python/cpython/blob/3.5/Lib/importlib/_bootstrap.py#L938-L962
+        # 3.6: https://github.com/python/cpython/blob/3.6/Lib/importlib/_bootstrap.py#L936-L960
+        # 3.7: https://github.com/python/cpython/blob/3.7/Lib/importlib/_bootstrap.py#L948-L972
+        # 3.8: https://github.com/python/cpython/blob/3.8/Lib/importlib/_bootstrap.py#L956-L980
+        # DEV: If we ever experience issues here we can always look at the code that executes/builds
+        #      an imported module instead since that should always get called when it is loaded
+        wrapt.wrap_function_wrapper('importlib._bootstrap', '_find_and_load_unlocked', wrapped_find_and_load_unlocked)
+
+        # 3.4: https://github.com/python/cpython/blob/3.4/Lib/importlib/__init__.py#L115-L156
+        # 3.5: https://github.com/python/cpython/blob/3.5/Lib/importlib/__init__.py#L132-L173
+        # 3.6: https://github.com/python/cpython/blob/3.6/Lib/importlib/__init__.py#L132-L173
+        # 3.7: https://github.com/python/cpython/blob/3.7/Lib/importlib/__init__.py#L133-L176
+        # 3.8: https://github.com/python/cpython/blob/3.8/Lib/importlib/__init__.py#L133-L176
+        wrapt.wrap_function_wrapper('importlib', 'reload', wrapped_reload)
+
+    # 2.7
+    # DEV: Slightly more direct approach of patching `__import__` and `reload` functions
```

Comment:
Why are we using that approach only for 2.7? Do we prefer to patch `importlib` for Py3?

Curious if doing the `__import__` patching for all versions would be simpler/safer.

Reply:
Revisiting the code, yeah, can probably just patch `importlib.__import__`. Think I was trying to be clever to save having to check if the module already exists in `sys.modules`, but really not worth it compared with patching an private function.

Reply:
oh, I remember now... Python 3 has multiple possible entrypoints for loading a module, patching `_find_and_load_unlocked` is a common/required function for any of them.

So gives us one place to patch rather than having to worry about all other possible places.

I'll update the comment above in PY3 patching.